dot/sync/warp_sync.go

```diff
@@ -0,0 +1,298 @@
+func (w *WarpSyncStrategy) NextActions() ([]*SyncTask, error) {
+	w.startedAt = time.Now()
+
+	lastBlock, err := w.lastBlockHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	var task SyncTask
+	switch w.phase {
+	case WarpProof:
+		task = SyncTask{
+			request:      messages.NewWarpProofRequest(lastBlock.Hash()),
+			response:     &messages.WarpSyncProof{},
+			requestMaker: w.reqMaker,
+		}
+	case TargetBlock:
+		req := messages.NewBlockRequest(
+			*messages.NewFromBlock(lastBlock.Hash()),
+			1,
+			messages.RequestedDataHeader+
+				messages.RequestedDataBody+
+				messages.RequestedDataJustification,
+			messages.Ascending,
+		)
+		task = SyncTask{
+			request:      req,
+			response:     &messages.BlockResponseMessage{},
+			requestMaker: w.reqMaker,
+		}
+	}
+
+	return []*SyncTask{&task}, nil
+}
+
+// Process processes the results of the sync tasks, getting the best warp sync response and
+// Updating our block state
+func (w *WarpSyncStrategy) Process(results []*SyncTaskResult) (
+	done bool, repChanges []Change, bans []peer.ID, err error) {
+
+	switch w.phase {
+	case WarpProof:
+		var warpProofResult *network.WarpSyncVerificationResult
+
+		repChanges, bans, warpProofResult = w.validateWarpSyncResults(results)
+
+		if warpProofResult != nil {
+			if !warpProofResult.Completed {
+				logger.Debug("partial warp sync received")
+
+				// Partial warp proof
+				w.setId = warpProofResult.SetId
+				w.authorities = warpProofResult.AuthorityList
+				w.lastBlock = &warpProofResult.Header
+			} else {
+				logger.Debug("complete warp sync received")
+
+				w.phase = TargetBlock
+				w.lastBlock = &warpProofResult.Header
+			}
+		}
+
+	case TargetBlock:
+		logger.Debug("getting target block")
+
+		var validRes []RequestResponseData
+
+		// Reuse same validator than in fullsync
+		repChanges, bans, validRes = validateResults(results, w.badBlocks)
+
+		if len(validRes) > 0 && len(validRes[0].responseData) > 0 {
+			w.result = *validRes[0].responseData[0]
+			w.phase = Completed
+		}
+	}
+
+	return w.IsSynced(), repChanges, bans, nil
+}
+
+func (w *WarpSyncStrategy) validateWarpSyncResults(results []*SyncTaskResult) (
+	repChanges []Change, peersToBlock []peer.ID, result *network.WarpSyncVerificationResult) {
+
+	repChanges = make([]Change, 0)
+	peersToBlock = make([]peer.ID, 0)
+	bestProof := &messages.WarpSyncProof{}
+	var bestResult *network.WarpSyncVerificationResult
+
+	for _, result := range results {
+		switch response := result.response.(type) {
+		case *messages.WarpSyncProof:
+			if !result.completed {
+				continue
+			}
+
+			// If invalid warp sync proof, then we should block the peer and update its reputation
+			encodedProof, err := response.Encode()
+			if err != nil {
+				// This should never happen since the proof is already decoded without issues
+				panic("fail to encode warp proof")
+			}
+
+			// Best proof will be the finished proof or the proof with more fragments
```

Comment:
I guess this comment used to be above `if response.IsFinished || len(response.Proofs) > len(bestProof.Proofs)`.

IMHO it can be removed since that line very clearly expresses this condition.

Reply:
Yess! thanks for the suggestion